src/main/java/org/spongepowered/api/service/permission/PermissionService.java

```diff
@@ -29,82 +29,238 @@
 import java.util.Collection;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
 
 
 /**
- * Represents a provider for permissions. This is the interface that a
- * permissions plugin must implement to provide permissions for a user.
+ * Represents a provider of permissions data.
+ *
+ * <p>This is the interface that a permissions plugin must implement and
+ * register to provide permissions for users on the server. Subject related
+ * method calls on Players and Users are redirected to the active permission
+ * service.</p>
+ *
+ * <p>Implementations of PermissionService are expected to be thread-safe.</p>
+ *
+ * <p>Distinctions are made throughout the service between data which is
+ * currently available, and data which may exist, but is not currently
+ * loaded.</p>
+ *
+ * <p>Methods prefixed with "get" will typically only return data if it is
+ * already loaded into memory. Methods prefixed with "load" however will load
+ * data into memory if necessary.</p>
+ *
+ * <p>{@link CompletableFuture}s are used throughout the service in situations
+ * where it is possible that calls may take a considerable time to execute.
+ * {@link CompletableFuture#thenAcceptAsync(Consumer, Executor)} can be used to
+ * run a callback once the method has completed. Alternatively,
+ * {@link CompletableFuture#join()} can be used, but care should be taken to
+ * avoid blocking the server thread.</p>
  */
 public interface PermissionService extends ContextualService<Subject> {
 
+    /**
+     * The standard identifier for the collection which stores users.
+     */
     String SUBJECTS_USER = "user";
```

Comment:
I'm not sure on the purpose of catalogs, but I understand they're similar to enums?

In which case, it's important to know that the "standard" identifiers expressed as static vars in the PermissionService class are by no means a definitive list of collections. Merely, they are the ones which are used by Sponge.

Other plugins are free to "create" or load their own SubjectCollections for their own purposes.

As long as it's possible to register new catalog entries at runtime, I guess this would be ok, but I still don't really see the benefit.

(it also breaks away from the same method signature for loading Subjects, currently loading a SubjectCollection & Subject is a very similar process, and the methods are named and behave in similar ways)